test/functional/p2p_filter.py

```diff
@@ -0,0 +1,107 @@
+#!/usr/bin/env python3
+# Copyright (c) 2020 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""
+Test BIP 37
+"""
+
+from test_framework.messages import (
+    MSG_BLOCK,
+    MSG_FILTERED_BLOCK,
+    msg_getdata,
+    msg_filterload,
+)
+from test_framework.mininode import (
+    P2PInterface,
+)
+from test_framework.test_framework import BitcoinTestFramework
+
+
+class FilterNode(P2PInterface):
+    # This is a P2SH watch-only wallet
+    watch_script_pubkey = 'a914ffffffffffffffffffffffffffffffffffffffff87'
+    # The initial filter with just the above scriptPubKey added
+    watch_filter_init = msg_filterload(
+        data=
+        b'@\x00\x08\x00\x80\x00\x00 \x00\xc0\x00 \x04\x00\x08$\x00\x04\x80\x00\x00 \x00\x00\x00\x00\x80\x00\x00@\x00\x02@ \x00',
+        nHashFuncs=19,
+        nTweak=0,
+        nFlags=1,
+    )
+
+    def on_inv(self, message):
+        want = msg_getdata()
+        for i in message.inv:
+            # inv messages can only contain TX or BLOCK, so translate BLOCK to FILTERED_BLOCK
+            if i.type == MSG_BLOCK:
+                i.type = MSG_FILTERED_BLOCK
+            want.inv.append(i)
+        if len(want.inv):
+            self.send_message(want)
+
+    def on_merkleblock(self, message):
+        self.merkleblock_received = True
+
+    def on_tx(self, message):
+        self.tx_received = True
+
+
+class FilterTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.setup_clean_chain = False
+        self.num_nodes = 1
+        self.extra_args = [[
+            '-peerbloomfilters',
+            "-whitelist=noban@127.0.0.1",  # immediate tx relay
+        ]]
+
+    def skip_test_if_missing_module(self):
+        self.skip_if_no_wallet()
+
+    def run_test(self):
+        self.log.info('Add filtered P2P connection to the node')
+        filter_node = self.nodes[0].add_p2p_connection(FilterNode())
+        filter_node.send_message(filter_node.watch_filter_init)
+        filter_node.sync_with_ping()
+        filter_address = self.nodes[0].decodescript(filter_node.watch_script_pubkey)['addresses'][0]
+
+        self.log.info('Check that we receive merkleblock and tx if the filter matches a tx in a block')
+        filter_node.merkleblock_received = False
+        filter_node.tx_received = False
+        self.nodes[0].generatetoaddress(1, filter_address)
+        filter_node.sync_with_ping()  # wait for inv
+        filter_node.sync_with_ping()  # wait for merkleblock
```

Comment:
This looks fishy, can two consecutive `sync_with_ping()` calls actually help more for synchronization than just one? To my understanding, `sync_with_ping` is only helpful whenever a node is already processing a received message and we want to guarantee that the processing is finished. I don't see how you could "wait for messages" with that.
We neither know that at the start of the first `sync_with_ping()` call the expected _inv_ message from node0 has already arrived, nor that at the second `sync_with_ping()` call _merkleblock_ has arrived (or even getdata was received from node0 before). I had a similar problem while working on https://github.com/bitcoin/bitcoin/pull/17461 (see https://github.com/bitcoin/bitcoin/pull/17461#issuecomment-590772489) with an adequate but still not 100% satisfactory waiting solution.

To prove my point, the test fails without the immediate tx relay enabled (which should only serve as speed-up, but not as synchronization mechanism, in my humble opinion).

Reply:
Thanks, replaced with `filter_node.wait_for_tx(txid)`, which is a stricter check that also passes when immediate tx relay is disabled.